sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -128,10 +131,23 @@ class EquivalentExpressions {
   // There are some special expressions that we should not recurse into all of its children.
   //   1. CodegenFallback: it's children will not be used to generate code (call eval() instead)
   //   2. ConditionalExpression: use its children that will always be evaluated.
-  private def childrenToRecurse(expr: Expression): Seq[Expression] = expr match {
-    case _: CodegenFallback => Nil
-    case c: ConditionalExpression => c.alwaysEvaluatedInputs
-    case other => other.children
+  private def childrenToRecurse(expr: Expression): Seq[Expression] = {
+    val alwaysEvaluated = expr match {
+      case _: CodegenFallback => Nil
+      case c: ConditionalExpression => c.alwaysEvaluatedInputs
+      case other => other.children
+    }
+    if (shortcut) {
+      // The subexpression may not need to eval even if it appears more than once.
+      // e.g., `if(or(a, and(b, b)))`, the expression `b` would be skipped if `a` is true.
+      alwaysEvaluated.map {
+        case and: And => and.left
+        case or: Or => or.left
+        case other => other
+      }
```

Comment:
Hmm, if other expression (e.g. `or.right`) is not added into recursing list, how can we look into if it needs to be eliminated? If `or.left` is false, it will be evaluated, isn't?

Reply:
It is why we add a new config with disabled by default. We can not decide which subexpression would be evaluated before running. When enable this config, it assumes that the left child is a shotcut, then the righ child can be skipped whatever it contains common subexpression.

Reply:
Another idea is to make CSE more dynamic: only evaluate it if its first appearance needs to be evaluated. It can handle `ConditionalExpression` as well but is much harder to implement.

Reply:
It's a kind of lazy evaluation. I remember before that pr attempts https://github.com/apache/spark/pull/32977 , and  there is some issues @viirya memtioned https://github.com/apache/spark/pull/32977#pullrequestreview-690266902.
It seems the main issue is that, the method will go to large if we make each common subexpression evaluation lazy. Something like:
```
def common_subexpression_1() {
  if (isnull) {
    // evaluate
  } else {
    // return exists value
  }
}
```